onnxruntime/python/onnxruntime_pybind_state.cc

```diff
@@ -1772,34 +1773,50 @@ void InitArray() {
   })();
 }
 
-// static variable used to create inference session and training session.
-static std::unique_ptr<Environment> session_env;
+namespace {
 
-void InitializeEnv() {
-  auto initialize = [&]() {
+class EnvInitializer {
+ public:
+  EnvInitializer() {
     // Initialization of the module
     InitArray();
+    std::unique_ptr<Environment> env_ptr;
     Env::Default().GetTelemetryProvider().SetLanguageProjection(OrtLanguageProjection::ORT_PROJECTION_PYTHON);
     OrtPybindThrowIfError(Environment::Create(std::make_unique<LoggingManager>(
                                                   std::make_unique<CLogSink>(),
                                                   Severity::kWARNING, false, LoggingManager::InstanceType::Default,
                                                   &SessionObjectInitializer::default_logger_id),
-                                              session_env));
+                                              env_ptr));
+    session_env_ = std::shared_ptr<Environment>(env_ptr.release());
+    destroyed = false;
+  }
 
-    static bool initialized = false;
-    if (initialized) {
-      return;
-    }
-    initialized = true;
-  };
-  initialize();
-}
+  ~EnvInitializer() {
+    destroyed = true;
+  }
+
+  std::shared_ptr<Environment> Get() const {
+    return session_env_;
+  }
+
+  static bool destroyed;
```

Comment:
>destroyed [](http://example.com/codeflow?start=14&length=9)

Why this be public updatable? 
And why it is static? The entire class should be a singleton no?

Reply:
It is static because it should be accessible before the class is constructed, and it is public because it should be accessible from the Get() function. Presumably, I could make Get() a part of the class, but that would call for more refactorings.